### PR TITLE
[RFC] core: remove algo from crypto_hash_*()

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Removes the algo parameters from all crypto_hash_*() functions except
crypto_hash_alloc_ctx().

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

With this pull request I'm exploring removal of the algo parameter from most `crypto_hash_*()` functions. 
- It is somewhat inconvenient for the caller to supply this parameter again and again
- It is redundant if the implementation already keeps track of which functions are used for this particular algorithm

What do you think? Should we continue down this path and take care of `crypto_cipher_*()`, `crypto_mac_*()` and `crypto_authenc_*()` too?

### CODE REVIEW

**Comment by jforissier** [comment]
LGTM.

Shouldn't `algo` rather be passed to `crypto_hash_init()` instead? The idea being, perhaps `crypto_hash_alloc_ctx()` is not mandatory and a context my be allocated by some other means?
(/me checks the code) Hmmm... I see that `hash_get_ctx_size()` is static, so not possible.

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by jenswi-linaro** [comment]
I was tempted to rename `crypto_hash_init()` to `crypto_hash_reset()` or `crypto_hash_reinit()`, but it doesn't harmonize so well with `crypto_{cipher,mac,authen}_init()`.

While reviewing the mbedtls integration there was one comment that `crypto_*_alloc_ctx()` may need to allocate memory in several layers if there's hardware involved. So a static allocation of the context is not very future proof.

I assume the acks means that you think it's a good idea to change `crypto_{cipher,mac,authen}_*()` also in a later P-R.

File: core/lib/libtomcrypt/src/tee_ltc_provider.c [hunk 1]
```diff
@@ -308,9 +313,9 @@ static TEE_Result hash_get_ctx_size(uint32_t algo, size_t *size)
 
 TEE_Result crypto_hash_alloc_ctx(void **ctx_ret, uint32_t algo)
 {
-	TEE_Result res;
-	size_t ctx_size;
-	void *ctx;
+	TEE_Result res = TEE_SUCCESS;
+	size_t ctx_size = 0;
+	struct hash_ctx *ctx;
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/lib/libtomcrypt/src/tee_ltc_provider.c]: I'll fix.

**Comment by jforissier** [comment]
> While reviewing the mbedtls integration there was one comment that `crypto_*_alloc_ctx()` may need to allocate memory in several layers if there's hardware involved. So a static allocation of the context is not very future proof

You're right, now I remember, too.

> I assume the acks means that you think it's a good idea to change `crypto_{cipher,mac,authen}_*()` also in a later P-R

Exactly. Thanks.

**Comment by jenswi-linaro** [comment]
Commen addressed, tag applied

**Comment by jenswi-linaro** [comment]
Maybe we should wait a little with this now that https://github.com/OP-TEE/optee_os/pull/2439 and https://github.com/OP-TEE/optee_os/pull/2705 is moving again.

**Comment by jforissier** [comment]
Agreed, there's no rush.

**Comment by jforissier** [comment]
I think this could be resumed now?

**Comment by jenswi-linaro** [comment]
Agree, I'll rebase etc when I can spare the time.

**Comment by jenswi-linaro** [comment]
Rebased and added commits for cipher, mac and authenc too. I think this PR should wait until after the release.

**Comment by jenswi-linaro** [comment]
Force pushed to rebase on master

**Comment by etienne-lms** [comment]
I was to ack the 4 commits, yet CI found issues: `tee_rpmb_fs.c` needs to be updated regarding `crypto_mac_*()` changes. I've tested few configs, it seems to me these are the only missing updates.

**Comment by jenswi-linaro** [comment]
Fixed up `core: remove algo from crypto_mac_*() `

**Comment by jenswi-linaro** [comment]
Rebased on master

**Comment by jforissier** [comment]
`Acked-by: Jerome Forissier <jerome@forissier.org>`

**Comment by jenswi-linaro** [comment]
Tag applied

**Comment by jforissier** [comment]
@etienne-lms ping?

**Comment by etienne-lms** [comment]
okay, i'll check by today.

File: core/crypto/aes-cts.c [hunk 2]
```diff
@@ -55,38 +56,156 @@ static TEE_Result cts_init(struct crypto_cipher_ctx *ctx,
...
+	TEE_Result res;
+	int nb_blocks, len_last_block, block_size = 16;
+	uint8_t tmp_block[64], tmp2_block[64];
+
+	if (!last_block)
+		return tee_do_cipher_update(cbc_ctx, TEE_ALG_AES_CBC_NOPAD,
+					     mode, last_block, data, len, dst);
+
+	/* Compute the last block length and check constraints */
+	nb_blocks = ((len + block_size - 1) / block_size);
```
Comment (etienne-lms) [inline core/crypto/aes-cts.c]: could remove parentheses.

File: core/crypto/aes-cts.c [hunk 3]
```diff
@@ -55,38 +56,156 @@ static TEE_Result cts_init(struct crypto_cipher_ctx *ctx,
...
+ *	5. Truncate the plaintext to the length of the original
+ *	   ciphertext.
+ */
+static TEE_Result cbc_cts_update(void *cbc_ctx, void *ecb_ctx,
+				 TEE_OperationMode mode, bool last_block,
+				 const uint8_t *data, size_t len, uint8_t *dst)
+{
+	TEE_Result res;
+	int nb_blocks, len_last_block, block_size = 16;
+	uint8_t tmp_block[64], tmp2_block[64];
```
Comment (etienne-lms) [inline core/crypto/aes-cts.c]: could default init these:
```
	int nb_blocks = 0;
	int len_last_block = 0;
	int block_size = 16;
	uint8_t tmp_block[64] = { };
	uint8_t tmp2_block[64] = { };

**Comment by jenswi-linaro** [comment]
Rebased, tag applied. Added a new commit ("core: fix style issues in cbc_cts_update()") as suggested.

**Comment by jenswi-linaro** [comment]
Tag applied

**Pull request merged**